### PR TITLE
shramkoweb: add frog-game task

### PR DESCRIPTION
**Pull request by Shramkoweb**
_Complete the form below and remove this line._

### task name

[Demo](https://shramkoweb.github.io/frontend-nanodegree-arcade-game/) |
[Code base](https://github.com/Shramkoweb/frontend-nanodegree-arcade-game)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/shramkoweb/frog-game/app.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        this.player.y < this.y + PLAYER_CONF.HEIGHT &&
+        this.player.y + PLAYER_CONF.HEIGHT > this.y
+    ) {
+        alert('You lose!')
+        this.player.x = PLAYER_CONF.START_X;
+        this.player.y = PLAYER_CONF.START_Y;
+    }
+};
+
+Enemy.prototype.update = function (dt) {
```
Comment (babayK0) [inline submissions/shramkoweb/frog-game/app.js:53]: The update method should only do rendering. Better to move the collision check from this method

[Separation of concerns](https://en.wikipedia.org/wiki/Separation_of_concerns)


You have commented part of code in `engine.js` file 

```
function update(dt) {
        updateEntities(dt);
        // checkCollisions();
    }
```
So it better to implement colliisions check outside of update, make a function checkCollisions, where you will call a method with collision checking for all entities

File: submissions/shramkoweb/frog-game/app.js [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+const Player = function (x, y) {
+    Character.call(this, x, y, PLAYER_CONF.SPRITE);
+};
+
+Player.prototype = Object.create(Character.prototype);
+Player.prototype.update = function () {
+    if (this.y < 0) {
+        this.x = PLAYER_CONF.START_X;
+        this.y = PLAYER_CONF.START_Y;
+        alert('You win!')
```
Comment (babayK0) [inline submissions/shramkoweb/frog-game/app.js:72]: Looks like I won without reaching the edge. Fix this, please.

Screenshot from 2020-10-16 00-04-16

**Review by babayK0 — CHANGES_REQUESTED** [review]
Good job !
Thanks for your clean code :smile: 
But i have some questions to you.
Why not use classes instead of prototypes ? It will make your code more cleaner than you have now.

Try to avoid force pushes in the future, but in this case it was normal :wink: 
Screenshot from 2020-10-15 23-57-55

Also see my comments below)

**Comment by Shramkoweb** [comment]
@babayK0 

> Why not use classes instead of prototypes ? It will make your code more cleaner than you have now.

The video https://classroom.udacity.com/courses/ud015 talked mostly about prototypes. So I used them.
If I understand correctly, classes are the topic of the next lesson?

File: submissions/shramkoweb/frog-game/app.js (hunk 1 above)
Comment (Shramkoweb) [reply submissions/shramkoweb/frog-game/app.js:53]: I need create one more method and call my isCollision inside them?

Comment (babayK0) [reply submissions/shramkoweb/frog-game/app.js:53]: This is comment from `engine.js`

```
/* This function is called by main (our game loop) and itself calls all
     * of the functions which may need to update entity's data. Based on how
     * you implement your collision detection (when two entities occupy the
     * same space, for instance when your character should die), you may find
     * the need to add an additional function call here. For now, we've left
     * it commented out - you may or may not want to implement this
     * functionality this way (you could just implement collision detection
     * on the entities themselves within your app.js file).
     */
```

Main point of this is to separete different logic in two separate methods, and than implement checkCollisions() in `engine.js` similary as updateEntities().


So update method do only coordinate reset,increasing speed(for Enemy).

If you dont understand me, i am open to your questions)

Comment (Shramkoweb) [reply submissions/shramkoweb/frog-game/app.js:53]: thx

**Comment by babayK0** [comment]
> @babayK0
> 
> > Why not use classes instead of prototypes ? It will make your code more cleaner than you have now.
> 
> In videos https://classroom.udacity.com/courses/ud015 talked mostly about prototypes. So I used them.
> If I understand correctly, classes are the topic of the next lesson?

I think if you know es6 classes you can made this task using them. But it is optional.

**Comment by Shramkoweb** [comment]
@babayK0 fix game.
move checkCollisions to engine.js u can see in source codes https://github.com/Shramkoweb/frontend-nanodegree-arcade-game/commit/4ec1d37a99efaa77ba7b3cdabcfcc14ac37187ac

**Comment by babayK0** [comment]
@Shramkoweb thank's a lot :wink: 

You can improve couple moments. (If you want) You have got temp unused method in Player.

```
Player.prototype.update = function () {
    // temp unused
};
```
Why not to execute this part of code inside a method update ?   
```
    this.x = PLAYER_CONF.START_X;
    this.y = PLAYER_CONF.START_Y;
```

And you can made start point a little less than 0 for the Enemy to start moving from behind the canvas, this will make the game visually smoother

However, I will approve  this pull request, well done! :+1:

**Review by babayK0 — APPROVED** [review]

**Comment by Shramkoweb** [comment]
> @Shramkoweb thank's a lot 😉
> 
> You can improve couple moments. (If you want) You have got temp unused method in Player.
> 
> ```
> Player.prototype.update = function () {
>     // temp unused
> };
> ```
> 
> Why not to execute this part of code inside a method update ?
> 
> ```
>     this.x = PLAYER_CONF.START_X;
>     this.y = PLAYER_CONF.START_Y;
> ```
> 
> And you can made start point a little less than 0 for the Enemy to start moving from behind the canvas, this will make the game visually smoother
> 
> However, I will approve this pull request, well done! 👍

U want move all code to update?
because i need before update(rerender) player and inly after call checkWin.
If i change position in update method its revert my old bag (https://github.com/kottans/frontend-2021-homeworks/pull/4#discussion_r505857288 ).

 ```
    this.x = PLAYER_CONF.START_X;
    this.y = PLAYER_CONF.START_Y;
 ```

**Comment by babayK0** [comment]
@Shramkoweb. 
:thinking: you are right. It is better to made one more method (reset maybe)
And than call it in parts where you need to reset position of Player, i guess.

> U want move all code to update?
> I think can't do this because i need first update (rerender) player and only after call checkWin.
> If i change position in update method its revert my old bag ([#4 (comment)](https://github.com/kottans/frontend-2021-homeworks/pull/4#discussion_r505857288) ).
> 
> ```
>    this.x = PLAYER_CONF.START_X;
>    this.y = PLAYER_CONF.START_Y;
> ```

So improvements: 
```
const ENEMIES_CONF = {
    ...
    START_X: -101,
};
....
Enemy.prototype.isCollision = function () {
    if (
        this.player.x < this.x + PLAYER_CONF.WIDTH &&
        this.player.x + PLAYER_CONF.WIDTH > this.x &&
        this.player.y < this.y + PLAYER_CONF.HEIGHT &&
        this.player.y + PLAYER_CONF.HEIGHT > this.y
    ) {
        alert('You lose!')
        this.player.reset(); // here
    }
};

....

// i think it is better then unused method. Or delete it if it is not useful
Player.prototype.update = function () { 
    this.x = this.x;
    this.y = this.y;
};

....

Player.prototype.reset = function () {
    this.x = PLAYER_CONF.START_X;
    this.y = PLAYER_CONF.START_Y;
};

...

Player.prototype.checkWin = function () {
    if (player.y <= 0) {
        window.requestAnimationFrame(() => {
            const needRestart = confirm('You win!');
            if (needRestart) {
                this.reset()
            }
        })
    }
};

```

**Comment by Shramkoweb** [comment]
> @Shramkoweb.
> 🤔 you are right. It is better to made one more method (reset maybe)
> And than call it in parts where you need to reset position of Player, i guess.
> 
> > U want move all code to update?
> > I think can't do this because i need first update (rerender) player and only after call checkWin.
> > If i change position in update method its revert my old bag ([#4 (comment)](https://github.com/kottans/frontend-2021-homeworks/pull/4#discussion_r505857288) ).
> > ```
> >    this.x = PLAYER_CONF.START_X;
> >    this.y = PLAYER_CONF.START_Y;
> > ```
> 
> So improvements:
> 
> ```
> const ENEMIES_CONF = {
>     ...
>     START_X: -101,
> };
> 
> ...
> 
> Enemy.prototype.update = function (dt) {
>     if (this.x > ctx.canvas.width) {
>         this.x = ENEMIES_CONF.START_X; // here
>         this.speed = this.getRandomSpeed();
>     } else {
>         this.x += this.speed * dt;
>     }
> };
> ....
> 
> Enemy.prototype.isCollision = function () {
>     if (
>         this.player.x < this.x + PLAYER_CONF.WIDTH &&
>         this.player.x + PLAYER_CONF.WIDTH > this.x &&
>         this.player.y < this.y + PLAYER_CONF.HEIGHT &&
>         this.player.y + PLAYER_CONF.HEIGHT > this.y
>     ) {
>         alert('You lose!')
>         this.player.reset(); // here
>     }
> };
> 
> ....
> 
> // i think it is better then unused method. Or delete it if it is not useful
> Player.prototype.update = function () { 
>     this.x = this.x;
>     this.y = this.y;
> };
> 
> ....
> 
> Player.prototype.reset = function () {
>     this.x = PLAYER_CONF.START_X;
>     this.y = PLAYER_CONF.START_Y;
> };
> 
> ...
> 
> Player.prototype.checkWin = function () {
>     if (player.y <= 0) {
>         window.requestAnimationFrame(() => {
>             const needRestart = confirm('You win!');
>             if (needRestart) {
>                 this.reset()
>             }
>         })
>     }
> };
> ```

Remove unused method update =)

**Pull request merged**